fe2/js/app.jsx

```diff
@@ -0,0 +1,266 @@
+      view = (
+        <Loading/>
+      );
+    } else {
+      view = (
+        <span>
+          <div className="col-sm-6">
+            <UploadPanel
+              image="static/img/upload-logo-docker.png"
+              message="Pull a Docker image from DockerHub"
+              >
+              <UploadForm source="dockerhub" btnMessage="Pull Image" url="api/v0/images" />
+            </UploadPanel>
+          </div>
+          <div className="col-sm-6">
+            <UploadPanel
+              image="static/img/upload-logo-github.png"
+              message="Build a Docker image from a GitHub Repo"
+              >
+              <UploadForm source="github" disabled={true} btnMessage="Build Image" url="api/v0/images" />
+            </UploadPanel>
+          </div>
+        </span>
+      );
+    }
+    return (
+      <div className="row">
+        <div className="col-sm-8 col-sm-offset-2">
+
+          <Nav activeKey={3} />
+
+          { view }
+
+        </div>
+      </div>
+    )
+  }
+})
+
+var UploadPanel = React.createClass({
+  render: function() {
+    return (
+      <div>
+        {this.props.message}
+        <img className="upload-logo" src={this.props.image}></img>
+        {this.props.children}
+      </div>
+    )
+  }
+})
+
+var UploadForm = React.createClass({
+  handleSubmit: function(e) {
+    UIDispatcher.handleAction({
+      type: ActionTypes.IMAGES_UPLOAD_SUBMIT_FORM,
+      data: {
+        source: this.props.source,
+        image: this.refs.image.getDOMNode().value.trim(),
+        url: this.props.url
+      }
+    });
+    e.preventDefault();
+  },
+  render: function() {
+    return (
+      <form className="uploadForm" onSubmit={this.handleSubmit}>
+        <input
+          type="text"
+          ref="image"
+          className="form-control"
+          disabled={this.props.disabled}
+        />
+        <br/>
+        <Button
+          bsStyle="primary"
+          bsSize="large"
+          block
+          disabled={this.props.disabled}>{this.props.btnMessage}</Button>
+      </form>
+    )
+  }
+})
+
+var Loading = React.createClass({
+  render: function() {
+    return (
+      <div className="row">
+        loading...
+      </div>
+    )
+  }
+})
+
+var UIDispatcher = assign(new Dispatcher(), {
+  handleAction: function(action) {
+    var payload = {
+      action: action
+    };
+    this.dispatch(payload);
+  }
+});
+
+var UIConstants = module.exports = {
+  ActionTypes: keyMirror({
+    IMAGES_UPLOAD_SUBMIT_FORM: null,
+  }),
+};
+var ActionTypes = UIConstants.ActionTypes;
+
+var ImagesStore = assign({}, EventEmitter.prototype, {
+  emitChange: function() {
+    this.emit(CHANGE_EVENT);
+  },
+  addChangeListener: function(callback) {
+    this.on(CHANGE_EVENT, callback);
+  },
+  removeChangeListener: function(callback) {
+    this.removeListener(CHANGE_EVENT, callback);
+  },
+});
+var CHANGE_EVENT = 'change';
+
+ImagesStore.dispatchToken = UIDispatcher.register(function(payload) {
```

Comment:
@danmane I'd love to figure out how to cleanly interop between jsx, ts, and js. Lack of static types puts stress on the flux pattern... makes communication more error-prone than it ought to be.